File: src/main/java/com/bazel_diff/BazelRule.java
package com.bazel_diff;

import java.util.List;

import com.google.common.hash.Hasher;
import com.google.common.hash.Hashing;
import com.google.devtools.build.lib.query2.proto.proto2api.Build;
import java.util.stream.Collectors;

interface BazelRule {
    byte[] getDigest();
    List<String> getRuleInputList();
    String getName();
}

class BazelRuleImpl implements BazelRule {
    private Build.Rule rule;

    BazelRuleImpl(Build.Rule rule) {
        this.rule = rule;
    }

    @Override
    public byte[] getDigest() {
        Hasher hasher = Hashing.sha256().newHasher();
        hasher.putBytes(rule.getRuleClassBytes().toByteArray());
        hasher.putBytes(rule.getNameBytes().toByteArray());
        hasher.putBytes(rule.getSkylarkEnvironmentHashCodeBytes().toByteArray());
        for (Build.Attribute attribute : rule.getAttributeList()) {
            hasher.putBytes(attribute.toByteArray());
        }
        return hasher.hash().asBytes();
    }

    @Override
    public List<String> getRuleInputList() {
        return rule.getRuleInputList()
                   .stream()
                   .map(ruleInput -> transformRuleInput(ruleInput))
                   .collect(Collectors.toList());
    }

    @Override
    public String getName() {
        return rule.getName();
    }

    private String transformRuleInput(String ruleInput) {
        if (ruleInput.startsWith("@")) {
            String[] splitRule = ruleInput.split("//");
            if (splitRule.length == 2) {
                String externalRule = splitRule[0];
                externalRule = externalRule.replaceFirst("@", "");
                return String.format("//external:%s", externalRule);
            }
        }
        return ruleInput;
    }
}

File: src/main/kotlin/com/bazel_diff/bazel/BazelRule.kt
package com.bazel_diff.bazel

import com.bazel_diff.hash.safePutBytes
import com.bazel_diff.hash.sha256
import com.google.devtools.build.lib.query2.proto.proto2api.Build

class BazelRule(private val rule: Build.Rule) {
    val digest: ByteArray by lazy {
        sha256 {
            safePutBytes(rule.ruleClassBytes.toByteArray())
            safePutBytes(rule.nameBytes.toByteArray())
            safePutBytes(rule.skylarkEnvironmentHashCodeBytes.toByteArray())
            for (attribute in rule.attributeList) {
                safePutBytes(attribute.toByteArray())
            }
        }
    }
    val ruleInputList: List<String>
        get() = rule.ruleInputList.map { ruleInput: String -> transformRuleInput(ruleInput) }

    val name: String = rule.name

    private fun transformRuleInput(ruleInput: String): String {
        if (ruleInput.startsWith("@")) {
            val splitRule = ruleInput.split("//".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (splitRule.size == 2) {
                var externalRule = splitRule[0]
                externalRule = externalRule.replaceFirst("@".toRegex(), "")
                return String.format("//external:%s", externalRule)
            }
        }
        return ruleInput
    }
}
